mindsdb/integrations/handlers/byom_handler/byom_handler.py

```diff
@@ -1,28 +1,174 @@
+        # pred_df = pred_df.rename(columns={target: 'prediction'})
+        return pred_df
+
+    def create_engine(self, connection_args):
+        # check code and requirements
+
+        model_proxy = self._get_model_proxy()
+
+        try:
+            model_proxy.check()
+        except Exception as e:
+            # remove venv
+            model_proxy.remove_venv()
+
+            raise e
+
+
+class ModelWrapper:
+    def __init__(self, code, modules_str, engine_id):
+        self.code = code
+        modules = self.parse_requirements(modules_str)
+
+        self.env_path = None
+        self.prepare_env(modules, engine_id)
+
+    def prepare_env(self, modules, engine_id):
+        config = Config()
+
+        try:
+            import virtualenv
+
+            base_path = config.get('byom', {}).get('venv_path')
+            if base_path is None:
+                # create in root path
+                base_path = Path(config.paths['root']) / 'venvs'
+
+            self.env_path = base_path / f'env_{engine_id}'
+
+            self.python_path = self.env_path / 'bin' / 'python'
+
+            if self.env_path.exists():
+                # already exists. it means requirements are already installed
+                return
+
+            # create
+            virtualenv.cli_run(['-p', sys.executable, str(self.env_path)])
+            log.logger.info(f"Created new environment: {self.env_path}")
+
+            if len(modules) > 0:
+                self.install_modules(modules)
+
+        except Exception as e:
+            log.logger.info("Can't create virtual environment. venv module should be installed")
+
+            self.python_path = Path(sys.executable)
+
+            # try to install modules everytime
+            self.install_modules(modules)
+
+    def remove_venv(self):
+        if self.env_path is not None and self.env_path.exists():
+            shutil.rmtree(str(self.env_path))
+
+    def parse_requirements(self, requirements):
+        # get requirements from string
+        # they should be located at the top of the file, before code
+
+        pattern = '^[\w\\[\\]-]+[=!<>\s]*[\d\.]*[,=!<>\s]*[\d\.]*$'
+        modules = []
+        for line in requirements.split(b'\n'):
+            line = line.decode().strip()
+            if line:
+                if re.match(pattern, line):
+                    modules.append(line)
+                else:
+                    raise Exception(f'Wrong requirement: {line}')
+
+        is_pandas = any([m.lower().startswith('pandas') for m in modules])
+        if not is_pandas:
+            modules.append('pandas >=1.1.5,<=1.3.3')
+
+        # for dataframe serialization
+        modules.append('pyarrow==11.0.0')
+        return modules
+
+    def install_modules(self, modules):
+        # install in current environment using pip
+
+        pip_cmd = self.python_path.parent / 'pip'
+        for module in modules:
+            os.system(f'{pip_cmd} install {module}')
```

Comment:
Maybe add an error check to raise an exception if something goes wrong (e.g. the user specified some package version that is not available for his specific OS version).

Reply:
what do you think about replacing it to?:
`if os.system('...') != 0: raise`

Reply:
Yes, that should work I think 👍